src/sparse/linalg/ordering.rs

```diff
@@ -13,74 +13,223 @@ pub struct Ordering<I> {
     pub connected_parts: Vec<usize>,
 }
 
-pub fn cuthill_mckee<N, I, Iptr>(mat: CsMatViewI<N, I, Iptr>) -> Ordering<I>
+pub fn reverse_cuthill_mckee<N, I, Iptr>(
+    mat: CsMatViewI<N, I, Iptr>,
+) -> Ordering<I>
 where
     N: PartialEq,
     I: SpIndex,
     Iptr: SpIndex,
 {
     debug_assert!(is_symmetric(&mat));
     assert_eq!(mat.cols(), mat.rows());
+
     let nb_vertices = mat.cols();
+    let degrees = mat.degrees();
+    let max_neighbors = degrees.iter().max().cloned().unwrap_or(0);
+
+    // This is the 'working data', into which new neighboring, sorted vertices are inserted,
+    // the next vertex to process is popped from here.
     let mut deque = VecDeque::with_capacity(nb_vertices);
-    let max_neighbors = mat
-        .indptr()
-        .windows(2)
-        .map(|w| w[1] - w[0])
-        .max()
-        .unwrap_or(Iptr::zero());
-    let mut neighbors = Vec::with_capacity(max_neighbors.index());
-    let mut perm = Vec::with_capacity(nb_vertices);
+
+    // This are all new neighbors of the currently processed vertex, they are collected here
+    // to be sorted prior to being appended to 'deque'.
+    let mut neighbors = Vec::with_capacity(max_neighbors);
+
+    // Storing which vertices have already been visited.
     let mut visited = vec![false; nb_vertices];
-    let mut connected_parts = Vec::with_capacity(4);
-    connected_parts.push(0);
 
-    let degrees = mat.degrees();
+    // Delimeting connected components inside the permutation.
+    let mut connected_parts = Vec::with_capacity(nb_vertices / 16 + 1);
+
+    // The final permutation reducing the bandwidth of the given sparse matrix.
+    // Missed optimization: Work with MaybeUninit here.
+    let mut perm = vec![I::default(); nb_vertices];
+
+    for perm_index in 0..nb_vertices {
+        // Find the next index to process, choosing a new starting vertex if necessary.
+        let current_vertex = deque.pop_front().unwrap_or_else(|| {
+            // We found a new connected component. This number will be reverse-inverted later.
+            connected_parts.push(perm_index);
+            // Employ the George-Liu pseudoperipheral vertex finder to find a new starting vertex.
+            find_pseudoperipheral_vertex(&visited, &degrees, &mat).unwrap()
```

Comment:
I'd like to be sure this unwrap cannot fail. It looks like `find_pseudoperipheral_vertex` can only fail if all vertices are visited, so it looks like this is panic free. If that is indeed the case, I'd like to see a comment stating it.

Reply:
That is correct. The `unwrap_or_else` can only be triggered if there is still an undiscovered component. This means `find_pseudoperipheral_vertex` can not fail, and this `unwrap` not be triggered.

Reply:
I will have to work on the error handling here anyway, custom strategies probably can't be trusted like this.